### ISSUE TITLE
Inline comments disappear when the line they're on is modified

### ISSUE THREAD

**Issue by jcansdale** [github/VisualStudio#1563]
- GitHub Extension for Visual Studio version: `2.4.4`

When making changes in the editor based on an inline comment, the inline comment disappears as soon as the line it's on is modified.

This is unfortunate for a few reasons:
- While working on the change the user might want to re-read the comment thread
- After the change is complete the user might want to reply with a description of what they've done

Having comments disappear can also be useful:
- It removes stale and hopefully resolved comments from the margin

We need a way to distinguish different levels of completeness:
1. The line a comment is on hasn't been touched and a comment is fresh
2. There are local changes that have made a comment stale
3. Changes have been pushed that have made a comment stale

#### Related

- Improve tracking of lines with inline-comments (spike/experiment): #1032

**Comment by grokys** [comment]
One thing I wanted to try there: how would it feel if the comment remained on the same line, even after the line had changed? Something like:

1. try to  match the comment with a new line
2. if it matches, update the line number
3. if no match use the most recently found matching line from 2. and grey the glyph

I'm not sure how well this would work, but it'd be interesting to see...

**Comment by jcansdale** [comment]
@grokys Yup, definitely want to do something along these lines! Better than giving up and putting it at the bottom.

I'm thinking something like this:

1. Take the set of comments we currently show on the PR diff (comments that are fresh on `github.com`)
2. Display comments normally who's diff information matches (as we currently do)
3. Do more forgiving line matching [1] when we can't match the diff and grey the glyph

That way the user will be able to see:
1. Stuff that hasn't been touched and they might need to work on
2. Stuff they've changed locally and need to describe on the comment thread and push

[1] Something similar to the navigate editor line matching algo I'm using. It looks for a unique match between one of the context lines and a line in the document (it can use the `  ` or `+` lines in the hunk). Once it finds a match, it can use it to position the comment (using an offset).
